src/handlers/diff_header.rs

```diff
@@ -153,7 +155,72 @@ impl<'a> StateMachine<'a> {
         {
             self.painter.emit()?;
             self._handle_diff_header_header_line(self.source == Source::DiffUnified)?;
-            self.handled_diff_header_header_line_file_pair = self.current_file_pair.clone()
+            self.handled_diff_header_header_line_file_pair = self.current_file_pair.clone();
+        }
+        Ok(handled_line)
+    }
+
+    #[inline]
+    fn test_diff_header_file_operation_line(&self) -> bool {
+        (matches!(self.state, State::DiffHeader(_)) || self.source == Source::DiffUnified)
+            && (self.line.starts_with("deleted file mode ")
+                || self.line.starts_with("new file mode "))
+    }
+
+    /// Check for and handle the "deleted file ..."  line.
+    pub fn handle_diff_header_file_operation_line(&mut self) -> std::io::Result<bool> {
+        if !self.test_diff_header_file_operation_line() {
+            return Ok(false);
+        }
+        let mut handled_line = false;
+        let (_mode_info, file_event) = parse_diff_header_line(
+            &self.line,
+            self.source == Source::GitDiff,
+            if self.config.relative_paths {
+                self.config.cwd_relative_to_repo_root.as_deref()
+            } else {
+                None
+            },
+        );
+        let name = get_repeated_file_path_from_diff_line(&self.diff_line)
+            .unwrap_or_else(|| "".to_string());
+        match file_event {
+            FileEvent::Removed => {
+                // self.mode_info = mode_info;
+                self.minus_file = name;
+                self.plus_file = "/dev/null".into();
+                self.minus_file_event = FileEvent::Change;
+                self.plus_file_event = FileEvent::Change;
+                self.current_file_pair = Some((self.minus_file.clone(), self.plus_file.clone()));
+            }
+            FileEvent::Added => {
+                // self.mode_info = mode_info;
+                self.minus_file = "/dev/null".into();
+                self.plus_file = name;
+                self.minus_file_event = FileEvent::Change;
+                self.plus_file_event = FileEvent::Change;
+                self.current_file_pair = Some((self.minus_file.clone(), self.plus_file.clone()));
+            }
+            _ => (),
+        }
+
+        // In color_only mode, raw_line's structure shouldn't be changed.
+        // So it needs to avoid fn _handle_diff_header_header_line
+        // (it connects the plus_file and minus_file),
+        // and to call fn handle_generic_diff_header_header_line directly.
+        if self.config.color_only {
+            write_generic_diff_header_header_line(
+                &self.line,
+                &self.raw_line,
+                &mut self.painter,
+                &mut self.mode_info,
+                self.config,
```

Comment:
(It would be nice to avoid the 3x repetition of this. Doesn't block this PR though.)